### PR TITLE
Fix bad uses of 'useTheme' during SSR

### PR DESCRIPTION
**Pull request by oetherington**
The simplest way to handle this seems to just be to not use `useTheme` in these contexts.

I added a `requireCssVar` function that provides a css variable for a particular theme item that will respect changes to the theme on the client side, and won't directly read the wrong theme during SSR.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1203286431838161) by [Unito](https://www.unito.io)
┆Link To Task: https://app.asana.com/0/1201302964208280/1203286431838161

### CODE REVIEW

**Comment by jpaddison3** [comment]
Doing it with css variables like we should.

**Comment by jpaddison3** [comment]
You missed a usage in ExpandedTagsSearchHit and ExpandedSequencesSearchHit.

Screen Shot 2022-11-04 at 6 03 16 PM

File: packages/lesswrong/components/posts/PostsPage/PostsPodcastPlayer.tsx [hunk 1]
```diff
@@ -2,17 +2,16 @@ import React, { useEffect, useRef } from 'react';
...
 import { spotifyPodcastIcon } from '../../icons/SpotifyPodcastIcon';
-import { useConcreteThemeOptions } from '../../themes/useTheme';
-import classNames from 'classnames';
 import { useEventListener } from '../../hooks/useEventListener';
 import { useTracking } from '../../../lib/analyticsEvents';
+import { requireCssVar } from '../../../themes/cssVars';
+
+const playerOpacity = requireCssVar("palette", "embeddedPlayer", "opacity");
 
 const styles = (): JssStyles => ({
```
Comment (jpaddison3) [inline packages/lesswrong/components/posts/PostsPage/PostsPodcastPlayer.tsx:11]:
```suggestion
const styles = (theme: ThemeType): JssStyles => ({
```

File: packages/lesswrong/components/posts/PostsPage/PostsPodcastPlayer.tsx [hunk 2]
```diff
@@ -2,17 +2,16 @@ import React, { useEffect, useRef } from 'react';
...
   embeddedPlayer: {
-    marginBottom: '2px'
-  },
-  playerDarkMode: {
-    opacity: 0.85
+    marginBottom: '2px',
+    opacity: playerOpacity,
```
Comment (jpaddison3) [inline packages/lesswrong/components/posts/PostsPage/PostsPodcastPlayer.tsx:14]:
```suggestion
    opacity: theme.palette.embeddedPlayer.opacity,
```

File: packages/lesswrong/components/themes/useTheme.tsx [hunk 3]
```diff
@@ -16,6 +16,12 @@ type ThemeContextObj = {
 }
 export const ThemeContext = React.createContext<ThemeContextObj|null>(null);
 
+/**
+ * NOTE: The hooks in this file should NOT be used for dynamicly applying styles/colors/etc.
+ * to components as this will have undesired results during SSR where we may or may not know
+ * which theme to use if the user has their theme set to "auto". For this use case you should
+ * instead use `requireCssVar`.
+ */
```
[the remark this answers is not in the document]
Comment (jpaddison3) [reply packages/lesswrong/components/themes/useTheme.tsx:24]: Maybe mention that.

File: packages/lesswrong/themes/cssVars.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * During SSR, we may not know which theme is currently being used if the user has their theme
+ * set to "auto", which can cause issues for components which need to include some specific
+ * styles conditionally. This function provides a work-around for this problem by exporting
+ * particular values from the theme as CSS variables, which will gracefully handle theme changes
+ * on the client side.
+ *
+ * For instance, if we want to use the theme value `theme.palette.panelBackground.dim`, we can do:
+ * const background = requireCssVar("palette", "panelBackground", "dim");
+ * const MyComponent = () => <div style={{background}} />;
```
Comment (jpaddison3) [inline packages/lesswrong/themes/cssVars.ts:23]: Which will set the style to be:

```
background: var(--palette:panelBackground:dim);
```

Which will then be gotten from the style sheet, which is auto-theme-safe.

Comment (jpaddison3) [reply packages/lesswrong/themes/cssVars.ts:23]: ^ add above to docstring

**Review by jpaddison3 — APPROVED** [review]
Very nice

File: packages/lesswrong/components/posts/PostsPage/PostsPodcastPlayer.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (oetherington) [reply packages/lesswrong/components/posts/PostsPage/PostsPodcastPlayer.tsx:14]: Doh

File: packages/lesswrong/components/themes/useTheme.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (oetherington) [reply packages/lesswrong/components/themes/useTheme.tsx:24]: There are still some valid uses (vote buttons fetch a delay timeout and the image uploader has to send some colors to an iframe that doesn't SSR for instance), but I've updated the comment to be a bit scarier so people hopefully don't use this in the future.

**Pull request merged**